src/polyfills/intersection-observer.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * See https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver.
+ */
+
+
+/* DO NOT SUBMIT: example:
+var inob = new IntersectionObserver(function(records) {
+  console.log('inob: ', records);
+});
+inob.observe($0);
+*/
+
+
+/**
+ * @param {!Window} win
+ */
+export function install(win) {
+  // DO NOT SUBMIT: remove, testing.
+  delete win.IntersectionObserver;
+  delete win.IntersectionObserverEntry;
+
+  if (!win.IntersectionObserver) {
+    win.IntersectionObserver = /** @type {typeof IntersectionObserver} */ (IntersectionObserverStub);
+  }
+}
+
+/**
+ * @param {!Window} childWin
+ */
+export function installForChildWin(childWin) {
+  const parent = childWin.parent;
+  if (!childWin.IntersectionObserver &&
+      parent &&
+      parent.IntersectionObserver) {
+    childWin.IntersectionObserver = parent.IntersectionObserver;
+    childWin.IntersectionObserverEntry = parent.IntersectionObserverEntry;
+  }
+}
+
+/**
+ * @param {!Window} win
+ */
+export function shouldLoadPolyfill(win) {
+  return (
+    !win.IntersectionObserver ||
+    win.IntersectionObserver === IntersectionObserverStub ||
+    !win.IntersectionObserverEntry ||
+    !('isIntersecting' in win.IntersectionObserverEntry.prototype)
+  );
+}
+
+/**
+ * @param {!Window} win
+ * @param {function()} installer
+ */
+export function upgradePolyfill(win, installer) {
+  // Can't use the IntersectionObserverStub here directly since it's a separate
+  // instance deployed in v0.js vs the polyfill extension.
+  const Stub = /** @type {typeof IntersectionObserverStub} */ (win.IntersectionObserver);
+  if (!Stub || Stub.stubs_) {
+    delete win.IntersectionObserver;
+    delete win.IntersectionObserverEntry;
+    installer();
+    const Impl = win.IntersectionObserver;
+    const stubs = Stub.stubs_;
+    Stub.stubs_ = [];
+    if (stubs.length > 0) {
+      const microtask = Promise.resolve();
+      stubs.forEach(stub => {
+        microtask.then(() => stub.upgrade_(Impl));
```

Comment:
Ditto comment about not being able to use private properties here, since they may be mangled differently.

Reply:
I redid this part. Instead of instances, I'm just pushing callbacks. So, in this instance, it's no longer an issue.